### ISSUE TITLE
Gemma 3 4B IT QAT (4-bit) produces incoherent repeated tokens

### ISSUE THREAD

**Issue by ekryski** [ekryski/mlx-swift-lm#170]
## Summary

`mlx-community/gemma-3-4b-it-qat-4bit` (registered in `VLMModelFactory` as `gemma3_4B_qat_4bit`, `Libraries/MLXVLM/VLMModelFactory.swift:193`) generates incoherent output consisting of a single repeated non-ASCII codepoint (e.g. `"Myմմմմմմմմմմմմմմմմմմմմմմմմմմմմմմ"`).

The bf16 variant `mlx-community/gemma-3-4b-it-bf16` produces coherent output under the same prompt and generation parameters, so the regression is specific to the QAT 4-bit checkpoint loading / quantization path in `MLXVLM/Models/Gemma3.swift`.

## Reproduction

- Model: `mlx-community/gemma-3-4b-it-qat-4bit`
- Output format: standard, no tools required to reproduce (single shell tool was registered in the failing log but not invoked)
- Generation params: `temperature=0.7, top_p=0.95, min_p=0.05, top_k=0, max_tokens=8192, max_kv_size=16384`, no repetition / presence penalty, kv quant disabled (`kv_bits=null`)
- System prompt: ~4760 chars
- Prompt: `"What is your name?"`

Observed:

```
TTFT: 778 ms
Total: 1149.5 ms
Output tokens: 31
Tokens/sec: 27.0
Raw output: "Myմմմմմմմմմմմմմմմմմմմմմմմմմմմմմմ"
```

Same prompt with `gemma-3-4b-it-bf16` produces normal output.

## Suspected cause

`Gemma3.swift` weight-sanitization branches on a custom `quantizationConfig` and falls back to `(64, 4, .affine)` when not provided (line 432). It also has a hand-rolled `hasQuantizedWeights` check that requires `.scales`, `.biases`, and `uint32 .weight` (lines 461–471). The QAT-4bit checkpoint may:

1. ship a `quantization_config` whose `group_size` / `bits` / `mode` differs from the assumed `(64, 4, .affine)`, or
2. lack `biases` for some layers (asymmetric vs. symmetric QAT), causing `hasQuantizedWeights` to return false and leaving those layers as raw `uint32` weights interpreted as bf16, or
3. have a different per-layer naming convention for vision tower vs. language model that the `language_model.<path>.scales` lookup misses.

Any of these would yield decoded weights that are numerically near-zero / saturated, producing the single-token loop we see (model collapses to its highest-probability token regardless of input).

The `27 tok/s` decode rate is in the expected range for a 4B 4-bit on Apple Silicon, so the inference loop itself is running normally — the issue is the weights being decoded incorrectly.

## What to check

- [ ] Diff the on-disk `quantization_config` in `mlx-community/gemma-3-4b-it-qat-4bit`'s `config.json` against the `(64, 4, .affine)` fallback in `sanitize(weights:)`.
- [ ] Compare against how `mlx-lm` (Python) loads the same checkpoint — does it use a different group size or `mxfp4` scheme?
- [ ] Confirm whether the QAT checkpoint includes `.biases` keys for every quantized layer, or whether some layers are symmetrically quantized.
- [ ] Verify the lm_head quantization detection (`hasQuantizedLmHead`) on this specific checkpoint — if lm_head is mis-quantized, the logits collapse exactly the way we see.
- [ ] Try generating with `MLXLM` text-only path (Gemma3 also exists as a text model) to isolate whether this is VLM-multimodal-tower-related or pure-language-model-loading.

**Comment by ekryski** [comment]
## Resolution — auto-fixed on alpha post-spec-006

The single-token-loop output (`"Myմմմմմմմմմմմ..."`) **no longer reproduces on `alpha` post-spec-006** ([PR #166](https://github.com/ekryski/mlx-swift-lm/pull/166)). Three smoke runs against `mlx-community/gemma-3-4b-it-qat-4bit` on `alpha @ v3.32.1-alpha`:

| Method | Output | Tokens | Decode tok/s |
|---|---|---|---|
| `simple` (greedy, ~30-token prompt) | `"Hello! My name is Gemma. I can help you with a variety of tasks, including:"` | 60 | 82.4 |
| `summarization` (1024-token Great Gatsby prompt) | `"Here's a summary of the text provided:"` | 241 | 72.0 |
| `vision` (golden retriever image, "Describe the image in English") | `"Okay/Here's a description of the image:"` | 172 | 84.6 |

All coherent. The original repro condition (long system prompt + `"What is your name?"`) was not specifically retried, but with three diverse contexts producing coherent output, the underlying weight-decoding bug is gone — a single-token-loop is a "weights interpreted as garbage" failure mode that wouldn't fix itself across context shapes.

### What likely fixed it

Confirmed inspection: `mlx-community/gemma-3-4b-it-qat-4bit/model.safetensors` does include the expected MLX-quantized layer triplets (`.weight` U32 + `.scales` BF16 + `.biases` BF16) for both language and vision towers. The `model.safetensors.index.json` file is *stale relative to the actual safetensors* (only lists 883 keys vs 1636 in the file), but the safetensors loader reads the file directly so the index discrepancy doesn't affect loading.

The most plausible auto-fix path: spec 006 changed how `loadWeights` interacts with `BaseConfiguration.PerLayerQuantization`. The legacy `quantization` parameter was deprecated in favour of `perLayerQuantization`, and the call now flows through `effectivePerLayerQuantization.quantization(layer: path)` which falls back to the default `Quantization(64, 4, .affine)` for layers without per-layer overrides. That path is exercised via the top-level `quantize(model:)` block in [`Load.swift:46-57`](Libraries/MLXLMCommon/Load.swift), so every layer with `.scales` keys gets quantized — independent of whether `Gemma3.languageModel.sanitize`'s own `hasQuantizedLmHead` gate fires.

### What remains worth tightening (not fixed in this sprint)

The Gemma 3 VLM language-model `sanitize` ([Libraries/MLXVLM/Models/Gemma3.swift:418-456](Libraries/MLXVLM/Models/Gemma3.swift)) still has the old quantization path:

```swift
let hasQuantizedLmHead = hasQuantizedWeights(layerPath: "language_model.lm_head", in: weights)
if hasQuantizedLmHead {
    let q = quantizationConfig?.asTuple ?? (64, 4, .affine)
    quantize(model: self) { path, module in ... }
}
```

This was the suspected failure path in the original issue. With spec 006's top-level `quantize(model:)` covering all layers via `perLayerQuantization`, this in-`sanitize` block becomes redundant (and runs first, so by the time the top-level block runs the layers are already quantized — possibly double-wrapping). It works currently, but it's dead-codey and worth removing in the **WS-C Gemma 3 consolidation PR** which will lift the shared text-decoder pieces into `MLXLMCommon/Models/Gemma3.swift`.

Closing this issue. The user-impact failure mode is fixed; the dead code in `sanitize` will be cleaned up during the consolidation refactor.
